### PR TITLE
Rewrite mdns_lite to support client-side use, DNS bridging, better testing/debugging, and fix mcast bugs

### PR DESCRIPTION
**Pull request by fhunleth**
I'm not expecting a proper review since this is a massive change. This is an API breaking change, but existing `config.exs` settings will continue to work. Luckily, most `mdns_lite` interactions are only via the `config.exs`.

The primary motivation for the change was to support client-side use of mDNS. I.e., it should be possible to connect to another Nerves device on the network via a `.local` name. This is now possible by using `MdnsLite.gethostbyname("hostname.local")`. 

However, making this change required some work since mDNS responders combine client and server functionality. It's not  good enough to just send an mDNS request to the mDNS multicast address/port. The async nature of mDNS means that you have to keep listening for mDNS advertisements. You can kind of fire a request on all network interfaces, collect responses for a little bit, and then stop, but then you're stuck doing only simple discovery and can't rely on any caching.

Once I started making then change, I switched the responders from going through a big block of logic to having them iterate through a list of DNS records to see which ones the query match and then use the RFC rules to run additional queries as an optimization. This change made it possible to have a functional core and made the unit tests easier. I made progress on some IPv6 support (AAAA records, etc.) since it was easy.

The next big change is a caching module. All records received on mDNS are added to a network interface-specific cache. The TTL's on the records are decremented and they're automatically removed. If the interface goes away, the whole cache goes away. The main table and the cache are queried before sending any multicast messages on the network. `MdnsLite.Info.dump_*` can be used to see what records are in built-in table and cache.

Since `MdnsLite.gethostbyname/1` can't be used by Erlang distribution and isn't used by the main Erlang name resolver, I added a DNS bridge. The DNS bridge is a simple DNS server that when it gets a `.local` name, it calls `MdnsLite.gethostbyname/1` and responds. Depending on the configuration, it will either error out or forward other names. Currently there's an Erlang/OTP issue (technically a Linux one) on handling errors so forwarding is the only viable option for now.

After this, I updated mDNS service advertisement handling. It was a little confusing before and I found myself misconfiguring on accident. I have a lot more checks now which I hope will be helpful. I removed the `:name` field since it looked important, but actually wasn't used except to identify services for removal at runtime. That field is now called `:id` and it's optional. The new `Options` module will normalize everything so old configs should do the "right" thing.

The docs should all be updated to describe the above. I feel like I need a fresh pass on the docs to make sure that they're in a cohesive since I've been updating them piecemeal for this change. At the moment, my energy level is such that I'm feeling like I need to get this in and then make smaller reviewable passes going forward.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,183 +4,200 @@
...
-  [
-    {:mdns_lite, "~> 0.6"}
-  ]
-end
+config :mdns_lite,
+  dns_bridge_enabled: true,
+  dns_bridge_recursive: true
+
+config :vintage_net,
+  additional_name_servers: [{127, 0, 0, 53}]
```
Comment (jjcarstens) [inline README.md:134]: Why is this needed and whats the logic for a `.53` address?

**Review by jjcarstens — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (fhunleth) [reply README.md:134]: It could be anything on 127.0.0.0/8. I picked 127.0.0.53 for a few reasons:

1. My first try was to just support a random port number, and I have partial support for that in `:vintage_net`, but then I found out that Erlang doesn't support non-port 53 specs in `resolv.conf`.
2. Second try was 127.0.0.1:53, but I didn't like that since it seemed some someone might run a DNS server on a Nerves device and it seemed better to have that port available.
3. The next try was picking a random local address. 127.0.0.53 seemed as good as any and that was the choice made by the Systemd people as well, so I did that.

Comment (fhunleth) [reply README.md:136]: I just added a little more text to the README on this.

**Review by jjcarstens — APPROVED** [review]
